[Documentation] Adding Developers Guide to docs

AlexandraLivadas opened the pull request:
This document offers general guidance to developers (both inside and outside the LORIS team) on how to contribute to LORIS and the general workflow of LORIS code development. It is still in progress. 

As of the latest commit, this doc was added to ReadTheDocs. To see a preview of how it renders, go to this link: https://loris-fork.readthedocs.io/en/2020-08-13-developersguidemarkdown/docs/wiki/99_Developers/Developers_Guide.html

TODO:

- [x] Complete the Code Review section

- [x] Add more detail to the Accessing a Database section

- [x] Fix formatting and wording

AlexandraLivadas commented on docs/wiki/99_Developers/Developers_Guide.md:
```diff
@@ -0,0 +1,230 @@
+        git cherry-pick <commit-hash>    # Cherry
+        git push -f origin "your-branch"
+        
+## 3. Making a Pull Request
```
This section is quite word-heavy and can be cut down.

AlexandraLivadas replied:
I cut down and re-formatted this to make it easier to read

christinerogers commented on docs/wiki/99_Developers/Developers_Guide.md:
```diff
@@ -0,0 +1,260 @@
+Don’t forget to make your branch names short and informative, like your commit messages, since they’re read by others on GitHub. 
+
+### Helpful commands to run before pushing new code:
+These commands catch small formatting erros that will cause Travis to fail.
```
typo - can you give the doc a proofread when you have a moment?

christinerogers commented on docs/wiki/99_Developers/Developers_Guide.md:
```diff
@@ -0,0 +1,260 @@
+### Tips for fixing a rebase
+1. The most basic way of “fixing” a rebase is to go back to the commit right before the rebase. The easiest way to do this is to either reset or revert to a previous commit. For example:
+    
+       git reset --hard <commit_hash>
```
some inconsistent indentation

christinerogers commented on docs/wiki/99_Developers/Developers_Guide.md:
```diff
@@ -0,0 +1,260 @@
+
+**[Contents](../../../.github/PULL_REQUEST_TEMPLATE.md) of initial comment:** Should include a brief summary of the changes and rationale, links to related issues or PRs, and testing instructions. 
+
+**Linking issues:** Use GitHub keywords like `Fixes` or `Resolves` so that the issue is closed automatically when the PR is merged.
```
add : do not use these key words if you are only addressing one point raised in the issue.  this will close the entire issue and leave unresolved bugs.

christinerogers commented on docs/wiki/99_Developers/Developers_Guide.md:
```diff
@@ -0,0 +1,260 @@
+To preview very large tables, use `\G` instead of `;` at the end of a query. This
+ formats the query output nicely, so that you can read through it easier. 
+
+
```
whitespace here should be removed in your next pass to clean up the typos, spacing, indentation, grammar, tone etc...

christinerogers commented on docs/wiki/99_Developers/Developers_Guide.md:
```diff
@@ -0,0 +1,260 @@
+* [Pull Request Checklist](https://github.com/aces/Loris/wiki/Code-Review-Checklist) and [Code Review guidelines](Code_Review_Guide.md)
+
+## 1. Working off your fork
+During installation, you should have already made your own fork of LORIS. Then, you should add two remote options so that you can switch between your fork and “aces”.
```
let's try to find a way to phrase these docs without "should"

christinerogers commented on docs/wiki/99_Developers/Developers_Guide.md:
```diff
@@ -73,66 +74,66 @@ before any git push can be done. To do this, run
 
 See also: [How to win the love of Travis](https://docs.google.com/presentation/d/1FDYOiyv-2m_nf2JIhJs8YjUCoHjELNyMNrHuqErD6O0/edit#slide=id.p). 
-(2019 presentation by John Saigle).
+(a 2019 presentation by John Saigle).
```
this should be team-internal.

christinerogers commented on docs/wiki/99_Developers/Developers_Guide.md:
```diff
@@ -35,12 +36,12 @@ There is usually no need to fully update your fork to the newest aces/LORIS vers
 
-**About Branch naming:** Follow the convention most devs use: “date-title...”.   
-Starting with the date allows branches to sort/display in order, and also helps sort out any conflicts faster in the main Loris repo.
+**About Branch naming:** Follow the convention most devevelopers use: “date-title...”.   
```
please proofread.

AlexandraLivadas replied:
this is a funny typo haha

christinerogers commented on docs/wiki/99_Developers/Developers_Guide.md:
```diff
@@ -73,66 +74,66 @@ before any git push can be done. To do this, run
-2. If you rebase the wrong branch or your PR is displaying all the commits from the branch you rebased and displays multiple commits or changed files 
-that are not part of your PR:   
-Use git cherry-pick. This command allows you to pick specific commits to apply to a branch. Here is an example workflow:  
+2. If you rebase the wrong branch or your PR is displaying all the commits from the branch you rebased and displays multiple commits or changed files that are not part of your PR,
```
super important point, doesn't read smoothly.  Can you take another look?

christinerogers commented on docs/wiki/99_Developers/Developers_Guide.md:
```diff
@@ -143,34 +144,34 @@ Be sure to observe and follow team conventions for how PRs should be named and d
 
-**Linking issues:** Use GitHub keywords like `Fixes` or `Resolves` so that the issue is closed automatically when the PR is merged.
-**Note** that automatic issue-linking and issue-closing only works on the default branch.  If the PR is not on the default branch, you’ll need to manually close the issue after merge. 
+**Linking issues:** Use GitHub keywords like `Fixes` or `Resolves` so that related issues are closed automatically when the PR is merged. Do not use these key words if you are only addressing one point raised in an issue. This will close the entire issue and leave unresolved bugs.
```
```suggestion
**Linking issues:** Use GitHub keywords like `Fixes` or `Resolves` so that related issues are closed automatically when the PR is merged. Do not use these keywords if you are only addressing one point raised in an issue. This will close the entire issue and leave unresolved bugs in the codebase.
```

christinerogers commented on docs/wiki/99_Developers/Developers_Guide.md:
```diff
@@ -143,34 +144,34 @@ Be sure to observe and follow team conventions for how PRs should be named and d
 Be sure to observe and follow conventions for how Issues should be named and described and how to ping other people for input.
  
-Generally we don’t assign issues.  If someone seems like an obvious candidate to take on an issue, bring their attention to it via slack, and they will self-assign if they feel appropriate.  
+Generally we do not assign issues. If someone seems like an obvious candidate to take on an issue, bring their attention to it via slack, and they will self-assign if they feel appropriate.  
```
this is mcin-internal.

christinerogers commented on docs/wiki/99_Developers/Developers_Guide.md:
```diff
@@ -219,10 +220,10 @@ You may ask for someone to review your PR at your discretion. However, do not as
 
 If you have requested review on a PR and have not received feedback for a number of days, it is okay to either re-request their review or leave a comment
-on the PR pinging them with a short message. For example, "Hey @developer, could you please review this PR and let me know what you think!"
+on the PR pinging the developer with a short message. For example, "Hey @developer, could you please review this PR and let me know what you think!"
```
let's leave exclamation marks out of public docs. 
```suggestion
on the PR pinging the developer with a short message. For example, "Hey @developer, could you please review this PR and let me know what you think."
```

christinerogers commented on docs/wiki/99_Developers/Developers_Guide.md:
```diff
@@ -219,10 +220,10 @@ You may ask for someone to review your PR at your discretion. However, do not as
 
 #### Implementing suggestions or requested changes
-It is good practice to implement any requested changes as soon as possible and then re-request review from the developer who requested them. It is also recommended
+It is good practice to implement any requested changes as soon as possible and then re-request review from the developer. It is also recommended
```
```suggestion
It is good practice to implement any requested changes as soon as possible and then re-request review from the developer. It is also helps the team if you acknowledge each comment/request with a few words or just an emoji, so that anyone reading will know you've seen it.```

christinerogers commented on docs/wiki/99_Developers/Developers_Guide.md:
```diff
@@ -241,20 +242,4 @@ mysql -u user -h host database -p
-
-
-
+To preview very large tables, use `\G` instead of `;` at the end of a query. This formats the query output nicely, so that you can read through it easier. 
```
this is a specific mysql commandline tip - 
```suggestion
MySQL commandline tip: try using `\G` instead of `;` at the end of a query - this will print one column value per line, which makes the query output easier to read. 
```

christinerogers commented on docs/wiki/99_Developers/Developers_Guide.md:
```diff
@@ -184,19 +185,19 @@ First, make a remote of their fork:
 
-    git fetch alex
-    git checkout alex/2020-05-06-something
+    git fetch remote_name
```
is `remote_name` the standard way we refer to this across all the install docs?  consistency is super important here.

AlexandraLivadas replied:
There's no other place in the public docs on Github that mentions forks/remote.
In the latest commit, I made sure everything in this doc was phrased in the same way across sections ("my_fork", "branch_name", etc)

christinerogers commented on docs/wiki/99_Developers/Developers_Guide.md:
```diff
@@ -143,34 +144,34 @@ Be sure to observe and follow team conventions for how PRs should be named and d
 When you make an issue, there are 4 options: bug fix, feature request, security 
 vulnerability, or a blank format. You will probably either be making a bug fix or feature 
-request. In this case, there will be a format to follow.
+request. In these cases, there will be a format to follow.
```
```suggestion
request. In these cases, there will be template to fill out.
```

AlexandraLivadas commented:
@christinerogers Concerning your comment about re-ordering the sections, I have changed their order to what I think is most valuable for outside developers:
1. Working off your fork, 2. Issues, 3. PRs, 4. Code Review, 5. Rebasing, 6. DB

christinerogers commented on docs/wiki/99_Developers/Developers_Guide.md:
```diff
@@ -0,0 +1,241 @@
+# LORIS Developers Guide
+This document covers topics including making PRs, issues, and how to go through a normal LORIS workflow.
```
```suggestion
This document covers topics including making PRs, issues, and how to go through a normal LORIS workflow, when contirbuting to the LORIS repository. Many notes in this guide are also relevant to contributing code on a fork or instance of LORIS.  
```

christinerogers commented on docs/wiki/99_Developers/Developers_Guide.md:
```diff
@@ -0,0 +1,241 @@
+# LORIS Developers Guide
+This document covers topics including making PRs, issues, and how to go through a normal LORIS workflow when contributing to the LORIS repository. Many notes in this guide are also relevant to contributing code on a fork or instance of LORIS.
+
+## Links to know, read, and follow
```
```suggestion
## Key Links
```

christinerogers commented on docs/wiki/99_Developers/Developers_Guide.md:
```diff
@@ -0,0 +1,241 @@
+This document covers topics including making PRs, issues, and how to go through a normal LORIS workflow when contributing to the LORIS repository. Many notes in this guide are also relevant to contributing code on a fork or instance of LORIS.
+
+## Links to know, read, and follow
+* All markdowns in the Loris GitHub under [docs/](../..) 
```
```suggestion
Important standards and practices documented in our repo include: 
```

christinerogers commented on docs/wiki/99_Developers/Developers_Guide.md:
```diff
@@ -0,0 +1,241 @@
+Any development should be done on your fork of LORIS, and **not** on aces/LORIS. 
+Before starting any PR, be mindful of which branch you want to base off of (main vs release-specific branch).   
+
+You will make a different branch for each PR or change that you are making.  
```
```suggestion
You will make a different branch for each PR or change that you are making.  
In general, smaller pull requests / changes are easier to work with in our team review context and will flow more smoothly through the review process.  
```

AlexandraLivadas replied:
This information was added in the internal document but I will add it here as well. Thanks 👍

driusan replied:
This thread looks like it should be marked as resolved? (The paragraph is already included below..)

christinerogers commented on docs/wiki/99_Developers/Developers_Guide.md:
```diff
@@ -0,0 +1,241 @@
+Don’t forget to make your branch names short and informative, like your commit messages, since they’re read by others on GitHub. 
+
+### Helpful commands to run before pushing new code:
+These commands catch small formatting errors that will cause Travis to fail.
```
```suggestion
These commands catch small formatting errors that will cause Travis CI (integrated on our GitHub) to fail.
```

christinerogers replied:
@AlexandraLivadas both here and in the next section talk about not tripping up travis, but Travis is only introduced in the next section. Could you change the headers / intro sentence so that this flows better?

christinerogers commented on docs/wiki/99_Developers/Developers_Guide.md:
```diff
@@ -0,0 +1,241 @@
+# LORIS Developers Guide
+This document covers topics including making PRs, issues, and how to go through a normal LORIS workflow when contributing to the LORIS repository. Many notes in this guide are also relevant to contributing code on a fork or instance of LORIS.
```
```suggestion
This document covers topics including making PRs, issues, and how to go through a normal LORIS workflow when contributing to the LORIS repository. Many notes in this guide are also relevant to contributing code on a fork or instance of LORIS, and many practices described here will be familiar to experienced developers.
```

christinerogers commented on docs/wiki/99_Developers/Developers_Guide.md:
```diff
@@ -0,0 +1,241 @@
+You can restart Travis if it gives error messages that seem inappropriate, or if the build stalls (for longer than 1h).
+if you are pushing a last typo fix on a PR, it is okay to add `[skip ci]` in the commit message. It is not okay to do this if it's the first commit or if you are fixing more than documentation typos.
+
+**Documentation PRs:** If working specifically on markdown files or other documentation, be sure to check the actual look of the document once you have pushed your PR and fix any formatting errors. You can do this by going to “Files Changed” and viewing how your markdown file will actually appear in GitHub. Don’t forget to click all links. 
```
```suggestion
**Documentation PRs:** If working specifically on markdown files or other documentation, be sure to check the actual look of the document once you have pushed your PR and fix any formatting errors. You can do this by going to “Files Changed” and viewing how your markdown file will actually appear in GitHub. Don’t forget to click all links to make sure they won't be broken when the document is rendered. 
```

christinerogers commented on docs/wiki/99_Developers/Developers_Guide.md:
```diff
@@ -0,0 +1,241 @@
+    git push --force my_fork branch_name
+    # --force is often needed here :\
+       
+This is a specific case. For more information on rebasing, look at this [(2014) Git for LORIS cheat document](https://docs.google.com/document/d/1yPx72Z7kJtV3WAcyrwtrMOAtQztzJecVQyaLCJ4jdHg/edit#).
```
internal only please.

christinerogers commented on docs/wiki/99_Developers/Developers_Guide.md:
```diff
@@ -0,0 +1,241 @@
+
+#### When to resolve a review or comment thread
+If you have implemented a requested change from a review-comment thread, it is good practice to let the developer who requested the change
+resolve the thread. However, if a comment thread is simple or straightforward and has been resolved, feel free to resolve it.
```
```suggestion
resolve the thread, so that they get the chance to check your change met their expectations. However, if a comment thread is simple or straightforward (e.g. a typo), feel free to resolve it.
```

christinerogers commented on docs/wiki/99_Developers/Developers_Guide.md:
```diff
@@ -0,0 +1,241 @@
+It is good practice to implement any requested changes as soon as possible and then re-request review from the developer. It also helps the team if you acknowledge each comment or request with a few words or just an emoji, so that anyone reading will know that you've seen it.
+
+If you feel that the suggested or requested change is not necessary, feel free to respond in the comment section with a clear explanation of why you don't think it is necessary.
+Make sure to stay positive and respectful in all review or comment threads.
```
tone.

christinerogers commented on docs/wiki/99_Developers/Developers_Guide.md:
```diff
@@ -0,0 +1,241 @@
+If you have requested review on a PR and have not received feedback for a number of days, it is okay to either re-request their review or leave a comment
+on the PR pinging the developer with a short message. For example, "Hey @developer, could you please review this PR and let me know what you think."
+
+#### Implementing suggestions or requested changes
```
this whole section is internal only.  "If you feel that the suggested or requested change is not necessary" is not appropriate for a guide for external devs. "be positive and respectful" etc

christinerogers commented on docs/wiki/99_Developers/Developers_Guide.md:
```diff
@@ -0,0 +1,241 @@
+Other Labels can also be added by reviewers at their discretion. 
+
+### Code Review Etiquette
+The most important thing to remember during code review is to be respectful and positive in all correspondence with other developers. It is also good practice
```
not for externals.

christinerogers commented on docs/wiki/99_Developers/Developers_Guide.md:
```diff
@@ -0,0 +1,241 @@
+The most important thing to remember during code review is to be respectful and positive in all correspondence with other developers. It is also good practice
+to be on top of all comments or requests and to respond to every comment.
+
+#### When to ask for or remind someone to do a review
```
significant tone update needed if this is going to be published.  please move to internal doc.

AlexandraLivadas commented:
@christinerogers I implemented the changes you requested above and also read through the document to check for tone issues. 
I removed a majority of the "Code Review Etiquette" section but kept the "When to resolve a review or comment thread" content because I think that information is specific to how Loris handles code review. Let me know what you think!

christinerogers commented:
thanks @h-karim for reviewing

@laemtl @spell00 -- reminder could you please spend a few minutes reading/reviewing this? thanks !

laemtl commented on docs/wiki/99_Developers/Developers_Guide.md:
```diff
@@ -0,0 +1,215 @@
+Setting up an ssh key can also be useful to avoid the headache of entering your username and password on each push. 
+
+## 2. Issues
+Issues are used to bring the team's attention to any bugs or feature requests that you come across. When you make an issue, there are 4 options: bug fix, feature request, security 
```
Should we remove this line break?

AlexandraLivadas replied:
@laemtl Which line break exactly? The one before the "Issues" header?

laemtl replied:
Between security and vulnerability. 
![Screenshot from 2020-08-25 15-26-38](https://user-images.githubusercontent.com/32041423/91218696-73089e80-e6e7-11ea-972f-c0a9d0c5b799.png)

laemtl commented on docs/wiki/99_Developers/Developers_Guide.md:
```diff
@@ -0,0 +1,215 @@
+
+Follow the provided testing instructions to test their code on your VM. 
+
+If you made changes (with their permission), commit and push them to that branch:
```
Now we need to request permission to be able to push on someone else repository. Maybe that worth a mention?

AlexandraLivadas replied:
That's a good point. Altho this is for external devs so I'm not quite sure if its the same process for them... 
@christinerogers Do you think more needs to be said here?

h-karim commented on docs/wiki/99_Developers/Developers_Guide.md:
```diff
@@ -0,0 +1,215 @@
+
+**Travis:** The Travis build status and results are linked near the bottom of your PR in GitHub. 
+You can restart Travis if it gives error messages that seem inappropriate, or if the build stalls (for longer than 1h).
+if you are pushing a last typo fix on a PR, it is okay to add `[skip ci]` in the commit message. It is not okay to do this if it's the first commit or if you are fixing more than documentation typos.
```
I think adding it under the "Good practice for Travis" is a good idea. It would help looking up anything related to Travis easier in case someone is looking for a quick reference.

AlexandraLivadas replied:
Sounds good! Thanks

laemtl commented on docs/wiki/99_Developers/Developers_Guide.md:
```diff
@@ -0,0 +1,215 @@
+    (#) while (conflicts)  { 
+        // fix any conflicts
+        git add  filename
+        git rebase --continue
```
or git rebase --skip if changes were discarded

laemtl commented on docs/wiki/99_Developers/Developers_Guide.md:
```diff
@@ -0,0 +1,215 @@
+    # --force is often needed here :\
+
+### Tips for fixing a rebase
+1. The most basic way of “fixing” a rebase is to go back to the commit right before the rebase. The easiest way to do this is to either reset or revert to a previous commit. 
```
Why do we need to fix a rebase? I'm not sure I understand this section :)

AlexandraLivadas replied:
Sometimes rebases get messed up. I can change the title to something like "How to fix a rebase gone wrong" or something

laemtl replied:
What is suggested here is a destructive operation. I suggest to perform a backup first. And also not sure why would want to destroy commits after a rebase?

spell00 commented on docs/wiki/99_Developers/Developers_Guide.md:
```diff
@@ -0,0 +1,219 @@
+
+**Do not** review or approve your own PR. Your PR must be reviewed and approved by two other developers before it can be merged. 
+
+**Important:** Do not merge your own PR or any other PR. After a PR has been reviewed and approved, the code is merged by Dave, the repository maintainer. It will be merged eventually. If your PR is on a release branch and has been ready to merge for a few days, you can add a comment that it is ready to merge. 
```
Is it necessary to mention the repository maintainer by name? It might not be him forever. Also, people outside Loris won't know who Dave is and it is not obvious to know that driusan is Dave

AlexandraLivadas replied:
That's a very good point. I mention Rida in this doc as well as the Raisinbread manager, so I will remove his name as well. Thanks!

spell00 commented on docs/wiki/99_Developers/Developers_Guide.md:
```diff
@@ -0,0 +1,219 @@
+**Notes:** 
+1. Hooks can be skipped with the flag `--no-verify: git push fork --no-verify`.  
+Setting up an ssh key can also be useful to avoid the headache of entering your username and password on each push. 
+2. If you are pushing a last typo fix on a PR, it is okay to add `[skip ci]` in the commit message. This will make the commit skip the Travis build. Do not do this if 
```
I tried `[skip ci]` in one of my PRs (#6671) and it hanged saying `continuous-integration/travis-ci  Expected  —  Waiting for status to be reported (required)` in the checks section. I did not wait because I had another modification to make, but I was expecting an immediate `All checks have passed`, but it's as if it was waiting for checks to run, but it would never actually happen?

AlexandraLivadas replied:
Thanks for testing this. The mention of [skip ci] was removed in the latest commit!

driusan commented on docs/wiki/99_Developers/Developers_Guide.md:
```diff
@@ -0,0 +1,219 @@
+    git commit -m "Initial commit"
+    git push my_fork branch_name
+
+There is usually no need to fully update your fork to the newest aces/LORIS version, since you can work off whichever branch you just fetched and checked-out.
```
I don't know what this sentence is supposed to mean.

AlexandraLivadas replied:
This is referring to how most devs don't update their fork of LORIS, they just fetch aces, checkout the main branch, and then make their own branch from there. I can try to reword this

christinerogers replied:
clarification would be useful.  maybe just switch the order of the clauses and spell this out a little more.

driusan commented on docs/wiki/99_Developers/Developers_Guide.md:
```diff
@@ -0,0 +1,219 @@
+ **Never** push to aces/Loris. Your branch will be deleted. Always push to your fork, and then create the PR from GitHub.
+
+#### Branch Naming
+Follow the convention most LORIS developers use: “date-title...”.   
```
I don't think that's a convention that most LORIS developers use

christinerogers replied:
it's a convention we appreciate and ask new developers to follow, part of Rida's conventional wisdom. it's also useful for new members.

driusan commented on docs/wiki/99_Developers/Developers_Guide.md:
```diff
@@ -0,0 +1,219 @@
+in the LORIS directory to configure the project. Make sure `.git/hooks/pre-push` is executable.
+
+**Notes:** 
+1. Hooks can be skipped with the flag `--no-verify: git push fork --no-verify`.  
```
```suggestion
1. Hooks can be skipped with the flag `--no-verify`. ie. `git push --no-verify`.  
```

The way this command is currently quoted is confusing.

driusan commented on docs/wiki/99_Developers/Developers_Guide.md:
```diff
@@ -0,0 +1,219 @@
+**Notes:** 
+1. Hooks can be skipped with the flag `--no-verify: git push fork --no-verify`.  
+Setting up an ssh key can also be useful to avoid the headache of entering your username and password on each push. 
+2. If you are pushing a last typo fix on a PR, it is okay to add `[skip ci]` in the commit message. This will make the commit skip the Travis build. Do not do this if 
```
`[skip ci]` should never be used on LORIS. PRs that use it won't be merged.

driusan commented on docs/wiki/99_Developers/Developers_Guide.md:
```diff
@@ -0,0 +1,219 @@
+
+PRs must be reviewed according to this [checklist](https://github.com/aces/Loris/wiki/Code-Review-Checklist), and must be approved by two reviewers before they are merged. Don’t forget to think about documentation updates like the TestPlan, changelog, and developer guide updates -- these must accompany your PR.
+
+Make sure to note special rules for SQL patches. Note that changes to Raisinbread data do not need patches, because Rida manages it.
```
```suggestion
Make sure to note special rules for SQL patches. Note that changes to Raisinbread data do not need patches, because raisinbread is maintained by sourcing the project patches.
```

driusan commented on docs/wiki/99_Developers/Developers_Guide.md:
```diff
@@ -0,0 +1,219 @@
+
+Make sure to note special rules for SQL patches. Note that changes to Raisinbread data do not need patches, because Rida manages it.
+
+**Do not** review or approve your own PR. Your PR must be reviewed and approved by two other developers before it can be merged. 
```
This sentence is unnecessary.. GitHub won't let you approve or request changes on your own PR.

driusan commented on docs/wiki/99_Developers/Developers_Guide.md:
```diff
@@ -0,0 +1,219 @@
+
+**Do not** review or approve your own PR. Your PR must be reviewed and approved by two other developers before it can be merged. 
+
+**Important:** Do not merge your own PR or any other PR. After a PR has been reviewed and approved, the code is merged by Dave, the repository maintainer. It will be merged eventually. If your PR is on a release branch and has been ready to merge for a few days, you can add a comment that it is ready to merge. 
```
```suggestion
**Important:** Do not merge your own PR or any other PR. After a PR has been reviewed and approved by a senior member of the LORIS team. It will be merged eventually. If your PR is on a release branch and has been ready to merge for a few days, you can add a comment that it is ready to merge. 
```

driusan commented on docs/wiki/99_Developers/Developers_Guide.md:
```diff
@@ -0,0 +1,219 @@
+
+It is your responsibility to be on top of your PRs and any Issues that need follow-up.  This includes answering questions and comments, acknowledging and making change requests, and requesting re-review after you’ve addressed someone’s change requests. 
+
+This is important for the Team’s productivity -- it prevents PRs from lagging and prevents confusion as to why a PR has stalled or if you’ve seen someone’s comments.
```
This sentence should be part of the previous paragraph, not a separate paragraph.. otherwise it's not immediately obvious what "this" is referring to. (Also "the Team" isn't a proper noun)

driusan commented on docs/wiki/99_Developers/Developers_Guide.md:
```diff
@@ -0,0 +1,219 @@
+It is your responsibility to be on top of your PRs and any Issues that need follow-up.  This includes answering questions and comments, acknowledging and making change requests, and requesting re-review after you’ve addressed someone’s change requests. 
+
+This is important for the Team’s productivity -- it prevents PRs from lagging and prevents confusion as to why a PR has stalled or if you’ve seen someone’s comments.
+When issuing a PR, you can request a specific reviewer (this is not the same as assigning a PR to someone).  There is usually an obvious reviewer for a PR.  
```
```suggestion
When issuing a PR, you can request a specific reviewer (this is not the same as assigning a PR to someone) if you know you would like someone in particular to review a PR.
```

driusan commented on docs/wiki/99_Developers/Developers_Guide.md:
```diff
@@ -0,0 +1,219 @@
+**[Contents](../../../.github/PULL_REQUEST_TEMPLATE.md) of the initial comment:** Include a brief summary of the changes and rationale, links to related issues or PRs, and testing instructions if there are any. 
+
+**Linking issues:** Use GitHub keywords like `Fixes` or `Resolves` so that related issues are closed automatically when the PR is merged. Do not use these keywords if you are only addressing one point raised in an issue. This will close the entire issue and leave unresolved bugs in the codebase.
+**Note** that automatic issue-linking and issue-closing only works on the default branch.  If the PR is not on the default branch, you will need to manually close the issue after the PR is merged. 
```
```suggestion
**Note** that automatic issue-linking and issue-closing only works on the default (main) branch.  If the PR is not on the default branch, you will need to manually close the issue after the PR is merged. 
```

driusan commented on docs/wiki/99_Developers/Developers_Guide.md:
```diff
@@ -0,0 +1,219 @@
+
+**Travis:** The Travis build status and results are linked near the bottom of your PR in GitHub. 
+You can restart Travis if it gives error messages that seem inappropriate, or if the build stalls (for longer than 1h).
+As mentioned above, if you are pushing a last typo fix on a PR, it is okay to add `[skip ci]` in the commit message. It is not okay to do this if it's the first commit or if you are fixing more than documentation typos.
```
No it's not.

driusan commented on docs/wiki/99_Developers/Developers_Guide.md:
```diff
@@ -0,0 +1,219 @@
+
+Alternatively, to avoid being under the detached HEAD state, checkout and create a local branch with the command
+
+    git checkout -b branch their_fork/branch_name
```
you still need the `git fetch` command

driusan commented on docs/wiki/99_Developers/Developers_Guide.md:
```diff
@@ -0,0 +1,219 @@
+        // OR you can "skip" specific changes
+        git rebase --skip
+    # } loop back to (#) - fix any remaining conflicts.... until no more conflicts
+    git push --force my_fork branch_name
```
```suggestion
    git push --force-with-lease my_fork branch_name
```

driusan commented on docs/wiki/99_Developers/Developers_Guide.md:
```diff
@@ -0,0 +1,219 @@
+        git rebase --skip
+    # } loop back to (#) - fix any remaining conflicts.... until no more conflicts
+    git push --force my_fork branch_name
+    # --force is often needed here :\
```
force will always be needed after a rebase.. force modifies history..

driusan commented on docs/wiki/99_Developers/Developers_Guide.md:
```diff
@@ -0,0 +1,219 @@
+For example:
+    
+       git reset --hard <commit_hash>
+       git push -f my_fork branch_name
```
```suggestion
       git push --force-with-lease my_fork branch_name
```

driusan commented on docs/wiki/99_Developers/Developers_Guide.md:
```diff
@@ -0,0 +1,219 @@
+    git checkout -b "branch_name"     # Create a “new” branch with the same
+                                      # name as the original branch
+    git cherry-pick <commit-hash>     # Cherry-pick the commits that you would like to keep
+    git push -f my_fork "branch_name"
```
```suggestion
    git push --force-with-lease my_fork "branch_name"
```

laemtl commented:
Any plan on adding this document to our documentation on readthedocs?

AlexandraLivadas commented:
@christinerogers What do you think about Laetitia's comment about adding this to ReadTheDocs?

laemtl commented on docs/wiki/99_Developers/Developers_Guide.md:
```diff
@@ -0,0 +1,212 @@
+#### Branch Naming
+Many LORIS developers follow the convention: “date-title...”.   
+Starting with the date allows branches to sort and display in order, and also helps sort out any conflicts faster in the main Loris repository. Please keep branch names short and informative.
+
```
### Development mode
To activate the development mode, turn sandbox to 1 in project/config.xml:
`<dev>
<sandbox>1</sandbox>
</dev>`

And then recompile:

`npm run compile`

This will change several configuration to allow more verbose error reporting, useful when debugging.

AlexandraLivadas replied:
I can add this, thanks :)

AlexandraLivadas replied:
Added!

christinerogers commented:
> @christinerogers What do you think about @laemtl's comment about adding this to ReadTheDocs?

Laetitia is 200% right - it should be in rendered in RTD.  

@laemtl and @AlexandraLivadas -- Please loop in @haoweiqiu so he sees and can document for us internally what are the steps to create/update docs and render them in RTD.

laemtl commented on docs/wiki/99_Developers/Developers_Guide.md:
```diff
@@ -0,0 +1,231 @@
+
+**Title:** [module_name] Quick description   
+
+**[Contents](../../../.github/PULL_REQUEST_TEMPLATE.md) of the initial comment:** Include a brief summary of the changes and rationale, links to related issues or PRs, and testing instructions if there are any. 
```
Absolute link is required here : https://github.com/AlexandraLivadas/Loris/blob/master/.github/PULL_REQUEST_TEMPLATE.md, ottherwise RTD will throw a 404 (this .md is not included in RTD).

laemtl commented on docs/wiki/99_Developers/Developers_Guide.md:
```diff
@@ -0,0 +1,231 @@
+### Fixing a rebase gone wrong
+1. The most basic way of “fixing” a rebase gone wrong is to go back to the commit right before the rebase. The easiest way to do this is to either reset or revert to a previous commit. **Note** that this is a potentially destructive command, so be very cautious when using it. It is also recommended that you make a backup of your branch before performing these commands so that you don't lose any c ...
+For example:
+
```
Remove this line break, otherwise RTD will start another numbered list (resulting in two consecutive 1.)

laemtl commented on docs/wiki/99_Developers/Developers_Guide.md:
````diff
@@ -0,0 +1,231 @@
+git reset --hard <commit_hash>
+git push --force-with-lease my_fork branch_name
+```
+    
````
And this one

laemtl commented on docs/wiki/99_Developers/Developers_Guide.md:
````diff
@@ -0,0 +1,231 @@
+```
+    
+2. A more complicated tool that can be used to solve bigger rebasing errors is the command [`git cherry-pick`](https://git-scm.com/docs/git-cherry-pick). This command allows you to pick specific commits to apply to a branch. For example, if you rebase the wrong branch or your PR is incorrectly displaying all the commits or files from the branch you rebased, this tool will come in handy. Here is a ...
+
````
Ans also this one

laemtl commented:
The numbered list issue is still not fixed. One of this solutions should work though: https://stackoverflow.com/questions/18088955/markdown-continue-numbered-list/22138846#22138846

AlexandraLivadas commented:
@christinerogers @driusan Can you re-review with the updated RTD formatting? Here is the link to view the RTD view: https://loris-fork.readthedocs.io/en/2020-08-13-developersguidemarkdown/docs/wiki/99_Developers/Developers_Guide.html

christinerogers reviewed (changes requested):
still some touchups needed, including replacing wiki links that shouldn't be used.

laemtl commented:
I tried to clarify the section. To remove the wiki link to https://github.com/aces/Loris/wiki/Code-Review-Checklist, this page needs to be migrated first.

christinerogers commented:
@driusan This would be useful to merge for our new group of developers.  Would you have a few minutes to take another look? 
Thanks --

christinerogers commented:
currently blocking a task for our 2021 GSOC student, aces/Loris-MRI#562
this PR should be merged before he can begin.

cmadjar commented:
@AlexandraLivadas could you please rebase so that Travis can run on it? Thank you!

AlexandraLivadas commented:
@driusan @christinerogers This is ready for review. All links are working and I've read over the content carefully :)

ridz1208 commented on docs/wiki/99_Developers/Developers_Guide.md:
```diff
@@ -0,0 +1,228 @@
+    npm run tests:unit -- --filter CandidateTest::someTest
+    npm run tests:integration
+
+### Good practice for Travis CI
```
Travis CI is fortunately no longer with us. maybe remove references to it and replace with GitHub Actions or automated tests

ridz1208 commented on docs/wiki/99_Developers/Developers_Guide.md:
```diff
@@ -0,0 +1,228 @@
+**[Contents](https://github.com/aces/Loris/blob/main/.github/PULL_REQUEST_TEMPLATE.md) of the initial comment:** Include a brief summary of the changes and rationale, links to related issues or PRs, and testing instructions if there are any. 
+
+**Linking issues:** Use GitHub keywords like `Fixes` or `Resolves` so that related issues are closed automatically when the PR is merged. Do not use these keywords if you are only addressing one point raised in an issue. This will close the entire issue and leave unresolved bugs in the codebase.
+**Note** that automatic issue-linking and issue-closing only works on the default (main) branch.  If the PR is not on the default branch, you will need to manually close the issue after the PR is merged. 
```
formatting: the note is at the same "level" as the title above it, I think it should be indented to show that its a note for `Linking issues:`

ridz1208 commented on docs/wiki/99_Developers/Developers_Guide.md:
```diff
@@ -0,0 +1,228 @@
+**Reviewers/Assignees:** Reviewers are people that you would like to ping specifically, but are not the same as assignees.
+Be sure to consult the Contributing and Code Review guidelines. For example, 2 reviewers must approve the PR and any code change must be pulled and tested by a developer before a PR is marked as `Passed Manual Tests`. 
+
+**Travis:** The Travis build status and results are linked near the bottom of your PR in GitHub. 
+You can restart Travis if it gives error messages that seem inappropriate, or if the build stalls (for longer than 1h).
```
no travis, github actions

ridz1208 commented on docs/wiki/99_Developers/Developers_Guide.md:
````diff
@@ -0,0 +1,228 @@
+    git push --force-with-lease my_fork "branch_name"
+    ```
+
+## 6. Accessing a Database
+To access the database, use the following command with the credentials given to you and provide the password on prompt:  
+```
+mysql -u user -h host database -p
+```
+
+If you are new to the MySQL commandline, try using `\G` instead of `;` at the end of a query - this will print one column value per line, which makes the query output easier to read.
````
I really don't think it's our job to explain this in any way. and if iit was, this would not be the place for it. by the time you get to developing you must have installed LORIS already and there are 100 places where you would have already connected.

In general we can't write documentation in silos, this is part of a larger documentation and should not reiterate this kind of details

ridz1208 reviewed (changes requested):
Other than the requested changes.... I'm not a big fan of the rebasing explanations partly because it's not our duty, these are not LORIS specific rebasing instructions, and partly because these instructions vary greatly from one case to another and from one developer to another... but I wont block over this

AlexandraLivadas commented:
> Other than the requested changes.... I'm not a big fan of the rebasing explanations partly because it's not our duty, these are not LORIS specific rebasing instructions, and partly because these instructions vary greatly from one case to another and from one developer to another... but I wont block over this

Hi @ridz1208! The rebasing instructions were added since this document is meant to cover all topics related to a normal Loris workflow for contributing developers. I do see your point though that the rebasing tips provided can be found elsewhere and aren't Loris specific, so I see no issue in removing this section. @christinerogers Is this okay with you?

cmadjar commented:
@ridz1208 to re-review and reply @AlexandraLivadas

ridz1208 commented:
@cmadjar @driusan I just realized this is waiting for confirmation from @christinerogers

christinerogers commented:
> Hi @ridz1208! The rebasing instructions were added since this document is meant to cover all topics related to a normal Loris workflow for contributing developers. I do see your point though that the rebasing tips provided can be found elsewhere and aren't Loris specific, so I see no issue in removing this section. @christinerogers Is this okay with you?

Hi Alex -- Can you move these please to the Internal LORIS Cheat sheet?  Thanks -- likewise for anything the senior team wants to cut (or not publish), just move it to the Loris cheat sheet.   thanks!

ridz1208 commented:
@christinerogers @AlexandraLivadas i think thats a good idea

driusan reviewed (dismissed):
Minor tweaking. Otherwise looks good!

christinerogers commented:
@AlexandraLivadas take a look at dave's change requests and i'll approve when addressed

christinerogers reviewed (approved):
Thanks @AlexandraLivadas for the most recent updates and improvements. 
@driusan i think this is ready for your final re-review, it looks like all your comments were addressed.

christinerogers commented:
@cmadjar is it time to bump this for review?   This should be merged for the release.  (i can add it to the team agenda too)

cmadjar commented:
@christinerogers I think this is more a question for @driusan or @ridz1208. I am mainly the note taker when I assign on GitHub, lol

AlexandraLivadas commented:
@driusan This PR is ready for review :)

driusan commented on docs/wiki/99_Developers/Developers_Guide.md:
```diff
@@ -0,0 +1,178 @@
+This will change several configurations in your LORIS project to allow for more verbose error reporting, which is useful for debugging.
+
+### Helpful commands to run before pushing new code:
+Travis is our continuous integration which runs on every PR in GitHub. These commands can catch small formatting errors that will cause the Travis build to fail.
```
This doesn't seem to be updated yet. We don't use Travis anymore, we use GitHub Actions.

driusan reviewed (dismissed):
Still referencing Travis.

christinerogers commented:
oh whoa this PR is still open.  @driusan @ridz1208 could this make it into the next minor release ? I'll tag it 24 in case that helps.

christinerogers commented:
> Still referencing Travis.

Dave's change request is still outstanding - 
@AlexandraLivadas can you resolve this quickly or would it take time?

ridz1208 commented:
I removed references to travis and I'll merge this when it passes tests

AlexandraLivadas commented:
@driusan I just fixed the conflict! The mkdocs version in `readthedocs/requirements.txt` has been updated since this PR was made so the conflict was in that file.

ridz1208 commented:
@driusan lets get this in
